File: .local/lib/copilot_workflows/worktree.py
"""Per-agent detached git worktrees: serialized, idempotent, auto-cleaned."""
from __future__ import annotations

import os
import re
import shutil
import subprocess
import threading
from collections.abc import Callable
from urllib.parse import unquote, urlparse

from ._util import noop

_SAFE = re.compile(r"[^A-Za-z0-9._-]+")


def _repo_name(repo: str) -> str:
    repo = repo.rstrip("/")
    parsed = urlparse(repo)
    path = unquote(parsed.path if parsed.scheme else repo)
    if "/_git/" in path:
        name = path.rsplit("/_git/", 1)[1]
    else:
        name = os.path.basename(path)
    return _SAFE.sub("-", name.removesuffix(".git")).strip("-.") or "repo"


def _repo_key(repo: str) -> str:
    """Best-effort repo identity; normalizes common GitHub and Azure clone URL variants."""
    if os.path.exists(repo):
        return "file:" + os.path.realpath(repo)
    if repo.startswith("git@"):
        repo = "ssh://git@" + repo[4:].replace(":", "/", 1)
    parsed = urlparse(repo.rstrip("/").removesuffix(".git"))
    path = unquote(parsed.path).strip("/")
    if "/_git/" in path:
        prefix, name = path.rsplit("/_git/", 1)
        parts = prefix.split("/")
        project = parts[-1]
        if parsed.netloc.endswith(".visualstudio.com"):
            org = parsed.netloc.split(".visualstudio.com", 1)[0]
        elif parsed.netloc == "dev.azure.com" and parts:
            org = parts[0]
        else:
            org = parsed.netloc
        return ("ado:%s/%s/%s" % (org, project, name)).lower()
    return ("%s/%s" % (parsed.netloc, path)).lower()


def ensure_clone(repo: str, dest: str, log: Callable[..., None] = noop) -> str:
    """Clone ``repo`` (a local path or full clone URL) into ``dest`` once; return ``dest``.

    Host-agnostic — the caller passes whatever URL `git clone` understands (GitHub, ADO, GitLab,
    ssh, file). Blob-less so a worktree-only review checkout stays cheap; reused if present.
    """
    if os.path.isdir(os.path.join(dest, ".git")):
        origin = _git(["remote", "get-url", "origin"], cwd=dest, check=False)
        if origin and _repo_key(origin) != _repo_key(repo):
            raise RuntimeError(
                "existing clone %s has origin %s, expected %s" % (dest, origin, repo))
    else:
        os.makedirs(os.path.dirname(dest) or ".", exist_ok=True)
        _git(["clone", "--filter=blob:none", "--no-checkout", repo, dest], cwd=os.getcwd())
        log("  clone %s" % repo)
    return dest


def clone_path(repo: str, clone_dir: str) -> str:
    """Return the persistent clone path for ``repo`` under ``clone_dir``."""
    return os.path.join(os.path.expanduser(clone_dir), _repo_name(repo))


def _git_result(args: list[str], cwd: str, check: bool = True) -> subprocess.CompletedProcess:
    proc = subprocess.run(
        ["git"] + args, cwd=cwd,
        stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True,
        encoding="utf-8", errors="replace",
    )
    if check and proc.returncode != 0:
        raise RuntimeError(f"git {' '.join(args)} failed: {proc.stderr.strip()}")
    return proc


def _git(args: list[str], cwd: str, check: bool = True) -> str:
    proc = _git_result(args, cwd, check)
    return proc.stdout.strip()


def find_repo_root(start: str) -> str | None:
    try:
        return _git(["rev-parse", "--show-toplevel"], cwd=start)
    except Exception:
        return None


class WorktreeManager:
    def __init__(self, repo_root: str, base_dir: str,
                 logger: Callable[..., None] | None = None, base_ref: str = "HEAD",
                 fetch_remote: bool = True) -> None:
        self.repo_root = repo_root
        self.base_dir = base_dir
        self.base_ref = base_ref
        self.fetch_remote = fetch_remote
        self._log = logger or noop
        self._lock = threading.Lock()
        self._created: list[str] = []

    def create(self, name: str, base_ref: str | None = None, fetch_ref: str | None = None) -> str:
        safe = _SAFE.sub("-", name).strip("-.") or "wt"  # strip dots too: "." / ".." alias base_dir
        path = os.path.join(self.base_dir, safe)
        real_base = os.path.realpath(self.base_dir)
        real_path = os.path.realpath(path)
        if real_path == real_base or not real_path.startswith(real_base + os.sep):
            raise RuntimeError(
                f"unsafe worktree name {name!r} resolves outside the worktree base")
        ref = base_ref or self.base_ref
        with self._lock:
            if path in self._created:
                raise RuntimeError(
                    f"worktree {name!r} is already active — use a unique name per concurrent branch")
            os.makedirs(self.base_dir, exist_ok=True)
            # A ref not in the clone yet (e.g. a PR head on another repo): fetch it and pin
            # the worktree to that fetched commit. Lets one base clone serve every PR's worktree.
            if fetch_ref:
                if self.fetch_remote:
                    _git(["fetch", "--depth", "1", "origin", fetch_ref], cwd=self.repo_root)
                    ref = "FETCH_HEAD"
                else:
                    ref = fetch_ref
            # A real worktree has a `.git` file at its root. Reuse a valid leftover from a
            # crashed run, but rebuild crash debris that isn't a worktree (a bare/partial dir).
            if os.path.exists(path) and not os.path.exists(os.path.join(path, ".git")):
                _git(["worktree", "remove", "--force", path], cwd=self.repo_root, check=False)
                shutil.rmtree(path, ignore_errors=True)
            if not os.path.exists(path):
                _git(["worktree", "add", "--detach", path, ref], cwd=self.repo_root)
            self._created.append(path)
            self._log(f"  worktree + {os.path.basename(path)}")
        return path

    def remove(self, path: str) -> None:
        with self._lock:
            if not os.path.exists(path):
                if path in self._created:
                    self._created.remove(path)
                return
            proc = _git_result(["worktree", "remove", "--force", path],
                               cwd=self.repo_root, check=False)
            if proc.returncode == 0 or not os.path.exists(path):
                if path in self._created:
                    self._created.remove(path)
                self._log(f"  worktree - {os.path.basename(path)}")
            else:
                self._log(
                    f"  ! worktree remove failed for {os.path.basename(path)}: "
                    f"{proc.stderr.strip()}")

    def cleanup_all(self) -> None:
        for path in list(self._created):
            self.remove(path)
        _git(["worktree", "prune"], cwd=self.repo_root, check=False)

File: .copilot/extensions/cwf/worktree.mjs
/**
 * @module worktree
 *
 * Per-agent detached git worktrees — a JS port of `worktree.py`. Serialized (async mutex),
 * idempotent, and auto-cleaned. Supports isolated worktrees for the launch repo, another local
 * repo, or a remote clone (blob-less, with a normal `--no-checkout` fallback when the server rejects
 * filtering), plus fetching a specific ref (e.g. a PR head). Two behaviours differ from the Python
 * reference, per the migration plan: a **clone `--filter=blob:none` fallback**, and **dirty
 * worktrees are preserved** for inspection instead of force-removed.
 *
 * Git runs asynchronously (`spawn`) so a clone never blocks the extension's event loop.
 */
import { spawn } from "node:child_process";
import { existsSync, mkdirSync, rmSync, realpathSync, statSync } from "node:fs";
import { join, dirname, basename, resolve, sep } from "node:path";

const SAFE = /[^A-Za-z0-9._-]+/g;
/** Sanitized name segment (dots stripped so `.`/`..` can't alias). @param {string} s */
const sanitize = (s) => s.replace(SAFE, "-").replace(/^[-.]+|[-.]+$/g, "") || "wt";

/** A tiny async mutex serializing a manager's create/remove (mirrors the Python threading.Lock). */
class Mutex {
	#tail = Promise.resolve();
	/** @template T @param {() => Promise<T>} fn @returns {Promise<T>} */
	run(fn) {
		const result = this.#tail.then(fn);
		this.#tail = result.then(
			() => {},
			() => {},
		);
		return result;
	}
}

/** @typedef {{ code: number, stdout: string, stderr: string }} GitResult */

/** Run a git command asynchronously. @param {string[]} args @param {string} cwd @returns {Promise<GitResult>} */
function gitResult(args, cwd) {
	return new Promise((res) => {
		const child = spawn("git", args, { cwd, stdio: ["ignore", "pipe", "pipe"] });
		let stdout = "";
		let stderr = "";
		child.stdout.setEncoding("utf8").on("data", (c) => (stdout += c));
		child.stderr.setEncoding("utf8").on("data", (c) => (stderr += c));
		child.on("error", (e) => res({ code: 127, stdout: "", stderr: String(e?.message || e) }));
		child.on("close", (code) => res({ code: code ?? 1, stdout, stderr }));
	});
}

/** Run git, returning trimmed stdout; throws on failure when `check`. @param {string[]} args @param {string} cwd @param {boolean} [check] */
async function git(args, cwd, check = true) {
	const r = await gitResult(args, cwd);
	if (check && r.code !== 0) throw new Error(`git ${args.join(" ")} failed: ${r.stderr.trim()}`);
	return r.stdout.trim();
}

/** Directory name for a persistent clone of `repo`. @param {string} repo */
export function repoName(repo) {
	repo = repo.replace(/\/+$/, "");
	let path;
	try {
		path = decodeURIComponent(new URL(repo).pathname);
	} catch {
		path = repo;
	}
	const name = path.includes("/_git/") ? path.split("/_git/").at(-1) : basename(path);
	return sanitize((name ?? "").replace(/\.git$/, "")) || "repo";
}

/** Best-effort repo identity (normalizes GitHub/ADO variants). @param {string} repo */
export function repoKey(repo) {
	if (existsSync(repo)) return "file:" + realpathSync(repo);
	if (repo.startsWith("git@")) repo = "ssh://git@" + repo.slice(4).replace(":", "/");
	let netloc = "";
	let path = repo;
	try {
		const u = new URL(repo.replace(/\/+$/, "").replace(/\.git$/, ""));
		netloc = u.host;
		path = decodeURIComponent(u.pathname).replace(/^\/+|\/+$/g, "");
	} catch {
		// treat as a bare path
	}
	if (path.includes("/_git/")) {
		const [prefix, name] = path.split("/_git/");
		const parts = prefix.split("/");
		const project = parts.at(-1);
		let org;
		if (netloc.endsWith(".visualstudio.com")) org = netloc.split(".visualstudio.com")[0];
		else if (netloc === "dev.azure.com" && parts.length) org = parts[0];
		else org = netloc;
		return `ado:${org}/${project}/${name}`.toLowerCase();
	}
	return `${netloc}/${path}`.toLowerCase();
}

/** Persistent clone path for `repo` under `cloneDir`. @param {string} repo @param {string} cloneDir */
export function clonePath(repo, cloneDir) {
	return join(resolve(cloneDir.replace(/^~(?=\/)/, process.env.HOME ?? "~")), repoName(repo));
}

/** Detect the git repo root containing `start`, or null. @param {string} start @returns {Promise<string|null>} */
export async function findRepoRoot(start) {
	const r = await gitResult(["rev-parse", "--show-toplevel"], start);
	return r.code === 0 ? r.stdout.trim() : null;
}

/**
 * Clone `repo` into `dest` once (blob-less, no-checkout); reuse an existing clone after validating
 * its origin. Falls back to a normal `--no-checkout` clone if the server rejects `--filter`.
 * @param {string} repo @param {string} dest @param {(m: string) => void} [log]
 * @returns {Promise<string>}
 */
export async function ensureClone(repo, dest, log = () => {}) {
	const gitDir = join(dest, ".git");
	if (existsSync(gitDir) && statSync(gitDir).isDirectory()) {
		const origin = await git(["remote", "get-url", "origin"], dest, false);
		if (origin && repoKey(origin) !== repoKey(repo)) throw new Error(`existing clone ${dest} has origin ${origin}, expected ${repo}`);
		return dest;
	}
	mkdirSync(dirname(dest) || ".", { recursive: true });
	const filtered = await gitResult(["clone", "--filter=blob:none", "--no-checkout", repo, dest], process.cwd());
	if (filtered.code === 0) {
		log(`  clone ${repo}`);
		return dest;
	}
	// Server may reject partial-clone filtering — retry a normal no-checkout clone.
	log(`  clone --filter rejected, retrying without filter: ${filtered.stderr.trim().split("\n").at(-1)}`);
	await git(["clone", "--no-checkout", repo, dest], process.cwd());
	log(`  clone ${repo}`);
	return dest;
}

/** Creates/removes detached worktrees under a base dir; serialized + idempotent. */
export class WorktreeManager {
	repoRoot;
	baseDir;
	baseRef;
	fetchRemote;
	/** @type {string[]} */
	preservedDirty = [];
	#log;
	#lock = new Mutex();
	/** @type {string[]} */
	#created = [];

	/**
	 * @param {string} repoRoot
	 * @param {string} baseDir
	 * @param {{ logger?: (m: string) => void, baseRef?: string, fetchRemote?: boolean }} [opts]
	 */
	constructor(repoRoot, baseDir, { logger = () => {}, baseRef = "HEAD", fetchRemote = true } = {}) {
		this.repoRoot = repoRoot;
		this.baseDir = baseDir;
		this.baseRef = baseRef;
		this.fetchRemote = fetchRemote;
		this.#log = logger;
	}

	/**
	 * Create a detached worktree named `name` (unique per concurrent branch). `fetchRef` (e.g.
	 * `pull/7/head`) is fetched and materialized when set. Returns the worktree path.
	 * @param {string} name @param {string|null} [baseRef] @param {string|null} [fetchRef]
	 * @returns {Promise<string>}
	 */
	create(name, baseRef = null, fetchRef = null) {
		return this.#lock.run(async () => {
			const safe = sanitize(name);
			const path = join(this.baseDir, safe);
			mkdirSync(this.baseDir, { recursive: true });
			// Contain the worktree within the base via realpath (resolves symlinks), like Python.
			const realBase = realpathSync(this.baseDir);
			const realPath = existsSync(path) ? realpathSync(path) : join(realBase, safe);
			if (realPath === realBase || !realPath.startsWith(realBase + sep)) throw new Error(`unsafe worktree name ${JSON.stringify(name)} resolves outside the worktree base`);
			if (this.#created.includes(path)) throw new Error(`worktree ${JSON.stringify(name)} is already active — use a unique name per concurrent branch`);
			let ref = baseRef || this.baseRef;
			if (fetchRef) {
				if (this.fetchRemote) {
					await git(["fetch", "--depth", "1", "origin", fetchRef], this.repoRoot);
					ref = "FETCH_HEAD";
				} else {
					ref = fetchRef;
				}
			}
			// Reuse a valid leftover worktree from a crashed run; rebuild non-worktree debris.
			if (existsSync(path) && !existsSync(join(path, ".git"))) {
				await gitResult(["worktree", "remove", "--force", path], this.repoRoot);
				rmSync(path, { recursive: true, force: true });
			}
			if (!existsSync(path)) await git(["worktree", "add", "--detach", path, ref], this.repoRoot);
			this.#created.push(path);
			this.#log(`  worktree + ${basename(path)}`);
			return path;
		});
	}

	/** Whether the worktree has uncommitted changes. @param {string} path @returns {Promise<boolean>} */
	async isDirty(path) {
		const r = await gitResult(["status", "--porcelain"], path);
		return r.code === 0 && r.stdout.trim().length > 0;
	}

	/**
	 * Remove a worktree — but **preserve it if dirty** (don't delete user/subagent work).
	 * @param {string} path @returns {Promise<void>}
	 */
	remove(path) {
		return this.#lock.run(async () => {
			const drop = () => {
				const i = this.#created.indexOf(path);
				if (i >= 0) this.#created.splice(i, 1);
			};
			if (!existsSync(path)) return drop();
			if (await this.isDirty(path)) {
				if (!this.preservedDirty.includes(path)) this.preservedDirty.push(path);
				drop();
				this.#log(`  worktree ~ ${basename(path)} preserved (uncommitted changes)`);
				return;
			}
			const r = await gitResult(["worktree", "remove", "--force", path], this.repoRoot);
			if (r.code === 0 || !existsSync(path)) {
				drop();
				this.#log(`  worktree - ${basename(path)}`);
			} else {
				this.#log(`  ! worktree remove failed for ${basename(path)}: ${r.stderr.trim()}`);
			}
		});
	}

	/** Remove every created worktree (preserving dirty ones) and prune. @returns {Promise<string[]>} preserved dirty paths. */
	async cleanupAll() {
		for (const path of [...this.#created]) await this.remove(path);
		await gitResult(["worktree", "prune"], this.repoRoot);
		return this.preservedDirty;
	}
}

export { sanitize as _sanitize, SAFE as _SAFE };
